### DISCUSSION TITLE
Custom lexer for Python-like languages

### DISCUSSION THREAD

**Discussion by vlkv** [eclipse-langium/langium#1085] [Q&A] [answered]
Hello, 

I am trying to figure out how to use some more sophisticated lexer instead of simple declaration of several terminal regexps. More specifically, I want lexer to produce into the resulting token stream additional tokens INDENT and DEDENT (which will mark out beginning of scope and ending of scope correspondingly). I found that in pure Chevrotain this task is accomplished with a bunch of [JavaScript logic](https://github.com/chevrotain/chevrotain/blob/master/examples/lexer/python_indentation/python_indentation.js). But how to integrate such a logic (or similar) into Languim?

Thank you.

**Comment by msujew** [comment, marked as answer]
Hey @vlkv,

you can override the [`TokenBuilder` service](https://github.com/langium/langium/blob/main/packages/langium/src/parser/token-builder.ts) to implement the Chevrotain pattern matching function in Langium.

**Reply by aabounegm** [reply]
Update: the [IndentationAwareTokenBuilder](https://eclipse-langium.github.io/langium/classes/langium.IndentationAwareTokenBuilder.html) has been released in v3.2.0, and there is now [a recipe](https://langium.org/docs/recipes/lexing/indentation-sensitive-languages/) explaining how to use it.
